network/native_endpointclient_linux.go

```diff
@@ -0,0 +1,438 @@
+func (client *NativeEndpointClient) ConfigureContainerInterfacesAndRoutesImpl(epInfo *EndpointInfo) error {
+
+	if err := client.netUtilsClient.AssignIPToInterface(client.containerVethName, epInfo.IPAddresses); err != nil {
+		return newErrorNativeEndpointClient("Failed to assign IPs to container veth interface", err.Error())
+	}
+	// kernel subnet route auto added by above call must be removed
+	for _, ipAddr := range epInfo.IPAddresses {
+		_, ipnet, _ := net.ParseCIDR(ipAddr.String())
+		routeInfo := RouteInfo{
+			Dst:      *ipnet,
+			Scope:    netlink.RT_SCOPE_LINK,
+			Protocol: netlink.RTPROT_KERNEL,
+		}
+		if err := deleteRoutes(client.netlink, client.netioshim, client.containerVethName, []RouteInfo{routeInfo}); err != nil {
+			return newErrorNativeEndpointClient("Failed to remove kernel subnet route", err.Error())
+		}
+	}
+
+	if err := client.AddDefaultRoutes(client.containerVethName); err != nil {
+		return newErrorNativeEndpointClient("Failed Container NS add default routes", err.Error())
+	}
+	if err := client.AddDefaultArp(client.containerVethName, client.vnetMac.String()); err != nil {
+		return newErrorNativeEndpointClient("Failed Container NS add default arp", err.Error())
+	}
+	return nil
+}
+
+// Called from ConfigureContainerInterfacesAndRoutes, Namespace: Vnet
+func (client *NativeEndpointClient) ConfigureVnetInterfacesAndRoutesImpl(epInfo *EndpointInfo) error {
+
+	err := client.netlink.SetLinkState(loopbackIf, true)
+	if err != nil {
+		return newErrorNativeEndpointClient("Failed to set loopback link state to up", err.Error())
+	}
+
+	// Add route specifying which device the pod ip(s) are on
+	routeInfoList := client.GetVnetRoutes(epInfo.IPAddresses)
+
+	if err = client.AddDefaultRoutes(client.ethXVethName); err != nil {
+		return newErrorNativeEndpointClient("Failed vnet NS add default/gateway routes (indempotent)", err.Error())
+	}
+	if err = client.AddDefaultArp(client.ethXVethName, azureMac); err != nil {
+		return newErrorNativeEndpointClient("Failed vnet NS add default ARP entry (idempotent)", err.Error())
+	}
+	if err := addRoutes(client.netlink, client.netioshim, client.vnetVethName, routeInfoList); err != nil {
+		return newErrorNativeEndpointClient("Failed adding routes to vnet specific to this container", err.Error())
+	}
+	// Return to ConfigureContainerInterfacesAndRoutes
+	return err
+}
+
+// Helper that gets the routes in the vnet NS for a particular list of IP addresses
+// Example: 192.168.0.4 dev <device which connects to NS with that IP> proto static
+func (client *NativeEndpointClient) GetVnetRoutes(ipAddresses []net.IPNet) []RouteInfo {
+	var routeInfoList []RouteInfo
+	// Add route specifying which device the pod ip(s) are on
+	for _, ipAddr := range ipAddresses {
+		var (
+			routeInfo RouteInfo
+			ipNet     net.IPNet
+		)
+
+		if ipAddr.IP.To4() != nil {
+			ipNet = net.IPNet{IP: ipAddr.IP, Mask: net.CIDRMask(ipv4FullMask, ipv4Bits)}
```

Comment:
I think new code should use the `netip` package instead of `net`. The types and fns are a significant improvement.

Reply:
I am unable to use the `netip` package because I need to call `addRoutes`, which takes in a `[]RouteInfo` parameter. This type requires types from `net.IP` (which is what is being constructed in the code above). There does not appear to be an equivalent of `addRoutes` in the `netip` package so I need to use the `addRoutes` that uses `net`.

Reply:
ack